Rewrite of DerivedTypeArgumentsTransformation

reuterbal opened the pull request:
_Note: This is based on #50._

This PR contains a rewrite of the `DerivedTypeArgumentsTransformation`. It enhances the functionality with recursive derived type argument expansion, i.e. is also able to flatten them when derived type arguments are handed through to another subroutine call, in which then array members are used. This is done by traversing the call tree from leaves upwards and storing the original argument names as well as the performed expansions in the trafo_data. That allows to apply this transformation as a single pass.

Included are loads of tricks around the handling of special edge cases, such as elemental routines, typebound procedure calls, named arguments, subroutines renamed upon import and the likes.

reuterbal commented:
I have applied the changes discussed offline and rebased this a couple of times onto main. 

Note, that this includes the interprocedural annotation API for InlineCalls that has also been filed separately with #93 .

@mlange please have another look and point out conceptual flaws that you are seeing, and we can then bring this into a useable state.

mlange05 reviewed (approved):
Ok many thanks! This is a very comprehensive change and covers a lot of corner cases and introduces a lot of complexity respectively to this Transformation. Admittedly, I cannot claim to understand everything in detail, but the technical coverage is solid and broad as usual, so I'm very happy with it. 

I've left a few comments, but none of them strictly require changes. The increased complexity makes picking up details a bit tricky sometimes, though, so I was wondering if there are any "lessons learned" or proposed API changes that could make this more streamlined?

In any case, happy and GTG from me, once CI is happy and up-to-date.

reuterbal commented on transformations/transformations/derived_types.py:
```diff
@@ -28,148 +33,502 @@
+
+        # ...and invalidate cached properties if dependencies have changed...
+        if dependencies_updated and item:
+            item.clear_cached_property('imports')
```
Caching of these properties will disappear in the scheduler rewrite

reuterbal commented on transformations/transformations/derived_types.py:
```diff
@@ -28,148 +33,502 @@
+        other_symbols = set()
+        call_mapper = {}
+        for call in FindNodes(CallStatement).visit(routine.body):
+            if not call.not_active:
```
Haha, true. Will change to 
```python
if not call.not_active == not False
```

reuterbal commented on transformations/transformations/derived_types.py:
```diff
@@ -28,148 +33,502 @@
+                call_name = str(call.name)
+                if call_name in successors_data:
+                    # Set the new call signature on the IR node
+                    arguments, kwarguments, others = self.expand_call_arguments(call, successors_data[call_name])
```
Agreed, naming isn't particularly good. That's other symbols that come in via imports - need to double-check but maybe this can actually be removed...

reuterbal replied:
Yes, this wasn't necessary anymore

reuterbal commented on transformations/transformations/derived_types.py:
```diff
@@ -28,148 +33,502 @@
+                other_symbols.update(others)
+
+        if call_mapper:
+            routine.body = SubstituteExpressions(call_mapper).visit(routine.body)
```
Yes, that's something I noticed as well. A combined mapper might be something useful maybe.

reuterbal commented on transformations/transformations/derived_types.py:
```diff
@@ -28,148 +33,502 @@
+        }
+        successors_data = CaseInsensitiveDict(
+            (
+                renamed_import_map.get(child.scope_name, {}).get(child.local_name, child.local_name),
```
Yes, this is to work around renaming upon import, which leads to calls having different names locally than the successor items they correspond to.

reuterbal commented:
Thanks for the helpful remarks, this helped me to identify another leftover from the special casing around elementary routines (which I think should rather be fixed in the fortran than mopped up here). 

Also, discovered and fixed two serious flaws:
- Derived types or parameters may need to be imported after flattening, e.g. when a kind parameter appears for the first time in a scope or when a derived type ends up in a routine's call signature, where it was previously deeply nested. These imports are now added after processing the kernel
- Handling of expanded kwarguments was broken

I hope this addressed the outstanding issues and CI completes successfully.

reuterbal commented:
Regarding the API question: 

Much of the awkard back-bending comes from two situations:
1. "Rebasing" a (nested) derived type member onto a local variable instance of the outer derived type
2. `arg_iter`-style iteration of `(callee argname, caller arg)` tuples, but working around the fact that the callee's signature has already been updated and we need to handle the 1-to-many mapping the flattening introduces.

I can see 1 being potentially useful in a wider sense eventually but 2 seems very niche to this use case.

Two other tricks might be interesting beyond this PR:
- Building the successor's list, taking local renaming of succesor items into account, e.g.:
```
use some_mod, only: local_name => name_in_mod
...
call local_name
```
Here, the successor item will use `name_in_mod`, while all calls and local use will be under `local_name` - these need matching up.
- Further unifying the interface of `InlineCall` and `CallStatement`, as many inter-procedural transformations will have to be applied to both, using almost identical patterns.

mlange05 commented:
Ok, looks good, going in.
